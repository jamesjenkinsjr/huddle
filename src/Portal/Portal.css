/* h2 {
  font-size: 4.5vw!important;
}

p {
  font-size: 3vw!important;
} */

h2 {
  color: #83d95b;
  font-size: 24px;
  margin: 0;
}

.portal__container {
  width: 80%;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.portal__loading-heading {
  font-size: 24px;
}

.portal__message-list {
  list-style-type: none;
  max-height: 300px;
  overflow: auto;
  white-space: pre-line;
  width: 90%;
  margin: 10px auto;
  border: 2px solid #6dd93b;
  color: #d7d7d9;
}

.portal__message-list > li.portal__message-item:nth-child(even) {
  border-top: 1px solid #d7d7d9;
  border-bottom: 1px solid #d7d7d9;
}

.portal__message-list > li {
  width: 98%;
  margin: 0 auto;
  padding: 5px 0;
}
.portal__expiry-datetime {
  font-size: 16px;
  color: #d7d7d9;
  text-align: center;
  margin: 4px auto;
}
.portal__close-session, .portal__copy-button {
  background-color: #83d95b;
  color: #3e3740;
  cursor: pointer;
  border: 2px solid #6dd93b;
  border-radius: 10px;
  padding: 4px;
  margin: 0 4px;
}

.portal__copy-button:hover {
  background-color: #3e3740;
  color: #6dd93b;
}

.portal__empty-messages {
  width: 90%;
  text-align: center;
  color: #d7d7d9;
  margin: 24px auto;
  border: 2px solid #6dd93b;
  padding: 5px 0;
}

.portal__gate-form {
  color: #d7d7d9;
}

.portal__gate-form > label > input {
  margin-left: 5px;
}

.portal__error {
  text-align: center;
  color: red;
}

@media screen and (min-width: 768px) {
  .portal__container {
    flex-direction: unset;
    flex-wrap: wrap;
    justify-content: space-evenly;
    align-items: center;
  }
  .portal__title {
    margin: 0;
  }

  .portal__message-list {
    max-height: 450px;
  }

  .portal__gate-form {
    margin: 30% auto;
  }
}

@media screen and (min-width: 1024px) {
  .portal__heading-container {
    display: flex;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    margin: 12px 5%;
  }

  .portal__message-list-heading {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: flex-start;
  }

  .portal__title {
    font-size: 48px;
  }

  .portal__expiry-datetime {
    font-size: 18px;
  }

  .portal__close-session:hover {
    background-color: #3e3740;
    border-color: red;
    color: red;
  }

  .portal__empty-messages {
    align-self: flex-start;
    width: 55%;
    height: 500px;
    padding-top: 30px;
  }

  .portal__message-list {
    list-style-type: none;
    align-self: flex-start;
    margin-bottom: 10%;
    height: 500px;
    width: 55%;
  }

  .portal__container {
    display: flex;
    flex-direction: row;
  }

  .portal__gate-form {
    margin: 20% auto;
  }

  .portal__gate-form > label {
    font-size: 18px;
  }
  .portal__gate-form > label > input {
    padding: 6px;
  }
}
