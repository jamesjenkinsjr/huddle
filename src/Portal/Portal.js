import React from 'react'
import './Portal.css'
import Message from '../Message/Message'
import PortalAPIService from '../services/portal-api-service'
import MessageForm from '../MessageForm/MessageForm'

export default class Portal extends React.Component {
  state = {
    error: '',
    loading: false,
  }

  handleRenderPortal = id => {
    this.setState({
      loading: true,
    })
    PortalAPIService.getPortalByID(id)
      .then(portal => {
        if (!portal) {
          this.setState({
            error: 'Portal does not exist',
          })
        }
        this.props.handlePortal(portal)
      })
      .then(() => {
        this.handleRenderMessages(this.props.portal.id)
      })
      .catch(error =>
        this.setState({
          error: error.message,
        })
      )
  }
  handleRenderMessages = id => {
    PortalAPIService.getPortalMessages(id)
      .then(messages => {
        this.props.handleMessages(messages)
        this.setState({ loading: false })
      })
      .catch(error => this.setState({ error: error.message }))
  }

  componentDidMount() {
    this.handleRenderPortal(this.props.match.params.id)
  }
  render() {
    const messages = this.props.messages.map(message => (
      <Message
        key={message.id}
        author={message.author}
        content={message.content}
        create_timestamp={new Date(message.create_timestamp).toLocaleString()}
      />
    ))
    if (this.state.error === '' && this.state.loading) {
      return <p>Loading Huddle...</p>
    } else if (this.state.error === '') {
      return (
        <section>
          <h1>{this.props.portal.name}</h1>
          {this.props.messages.length > 0 && (
            <ul className="portal__message-list">{messages}</ul>
          )}
          {this.props.messages.length === 0 && <p>No messages found</p>}
          <MessageForm handleNewMessage={this.props.handleNewMessage} portal_id={this.props.match.params.id}/>
        </section>
      )
    } else {
      return <h1>{this.state.error}</h1>
    }
  }
}
