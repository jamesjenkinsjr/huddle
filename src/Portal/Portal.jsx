import React from 'react'
import './Portal.css'
import Message from '../Message/Message'
import PortalAPIService from '../services/portal-api-service'
import MessageForm from '../MessageForm/MessageForm'
import TokenService from '../services/token-service'

export default class Portal extends React.Component {
  static defaultProps = {
    portal: {},
    messages: [],
    match: {
      params: {
        id: '',
      },
    },
    history: {
      push: () => {},
    },
  }
  state = {
    error: '',
    loading: true,
    gated: false,
  }

  componentDidMount() {
    this.setState({
      loading: false,
    })

    this.handleRenderPortal(this.props.match.params.id)
    this.interval = setInterval(
      () => this.handleRenderMessages(this.props.portal.id),
      5000
    )
    if (this.scrollEl) {
      this.scrollToBottom()
    }
  }

  componentDidUpdate() {
    if (this.scrollEl) {
      this.scrollToBottom()
    }
  }

  componentWillUnmount() {
    clearInterval(this.interval)
  }

  scrollToBottom = () => {
    this.scrollEl.scrollIntoView({ behavior: 'auto' })
  }

  handleRenderPortal = id => {
    this.setState({
      loading: true,
    })
    PortalAPIService.getPortalByID(id)
      .then(portal => {
        if (!portal) {
          this.setState({
            error: 'Portal does not exist',
          })
        }
        this.props.handlePortal(portal)
      })
      .then(() => {
        /**
         * Check if portal does not require password before
         * loading portal messages. If it does, render
         * check that an auth token is set before attempting
         * intial message loading
         */
        if (
          !this.props.portal.use_password ||
          TokenService.hasPortalToken(this.props.portal.id)
        ) {
          this.handleRenderMessages(this.props.portal.id)
        }
      })
      .then(() => {
        this.setState({
          loading: false,
          gated: this.props.portal.use_password,
        })
      })
      .catch(error => {
        if (error.message === 'Unauthorized portal request') {
          this.setState({
            error: error.message,
            gated: true,
          })
        } else {
          this.setState({
            error: error.message,
          })
        }
      })
  }

  handleRenderMessages = id => {
    if (!id) {
      return
    }
    if (
      /**
       * Depending on status of portal validation,
       * check for both props or local state to police
       * messages loading.
       */

      (this.props.portal.use_password || this.state.gated) &&
      !TokenService.hasPortalToken(this.props.match.params.id)
    ) {
      return
    }
    PortalAPIService.getPortalMessages(id)
      .then(messages => {
        if (messages.length !== this.props.messages.length) {
          this.props.handleMessages(messages)
          this.setState({ loading: false })
        }
      })
      .catch(error => this.setState({ error: error.message, loading: false }))
  }

  handlePortalValidation = e => {
    e.preventDefault()
    const password = e.target.validate_password.value
    e.target.validate_password.value = ''
    PortalAPIService.authorizeGatedPortal(this.props.match.params.id, {
      password,
    })
      .then(res => {
        TokenService.setPortalToken(this.props.match.params.id, res.portalAuth)
        this.setState({
          validated: true,
          error: '',
        })
      })
      .then(() => this.handleRenderPortal(this.props.match.params.id))
      .catch(error => this.setState({ error: error.message }))
  }

  handleClosePortal = e => {
    e.preventDefault()
    if (this.state.gated) {
      TokenService.clearPortalToken(this.props.portal.id)
      this.handleRenderPortal(this.props.portal.id)
    } else {
      this.props.history.push('/')
    }
  }

  handleCopyURL = e => {
    const copyButton = document.getElementById('copy')
    // Chrome and Firefox allow for a much cleaner copy
    // to clipboard experience
    if (!!navigator.clipboard) {
      navigator.clipboard
        .writeText(e.target.value)
        .then(() => {
          copyButton.innerHTML = 'Copied Portal URL!'
          setTimeout(() => (copyButton.innerHTML = 'Copy Portal URL'), 2000)
        })
        .catch(err => {
          copyButton.innerHTML = 'Try Again'
          setTimeout(() => (copyButton.innerHTML = 'Copy Portal URL'), 2000)
        })
    } else {
      // for Safari and iOS devices we leverage an off-screen
      // input that we can store the URL in and select and copy from
      const input = document.getElementById('copy_data')
      input.setAttribute('value', e.target.value)
      const el = input
      const editable = el.contentEditable
      const readOnly = el.readOnly
      const range = document.createRange()
      range.selectNodeContents(el)
      const selection = window.getSelection()
      selection.removeAllRanges()
      selection.addRange(range)
      el.focus()
      el.setSelectionRange(0, 999999)
      el.contentEditable = editable
      el.readOnly = readOnly
      document.execCommand('copy')

      copyButton.innerHTML = 'Copied Portal URL!'
      setTimeout(() => (copyButton.innerHTML = 'Copy Portal URL'), 2000)
    }
  }

  render() {
    const messages = this.props.messages.map(message => (
      <Message
        key={message.id}
        author={message.author}
        content={message.content}
        create_timestamp={new Date(message.create_timestamp).toLocaleString()}
      />
    ))
    if (
      !this.state.loading &&
      this.state.error === '' &&
      (!this.state.gated ||
        TokenService.hasPortalToken(this.props.match.params.id))
    ) {
      return (
        <section className="portal__container">
          <div className="portal__heading-container">
            <div className="portal__message-list-heading">
              <h2 className="portal__title">{this.props.portal.name}</h2>
              <p className="portal__expiry-datetime">
                Expires on{' '}
                {new Date(this.props.portal.expiry_timestamp).toLocaleString()}
              </p>
            </div>
            <div className="portal__heading-button-container">
              <input
                type="text"
                name="copy_data"
                id="copy_data"
                // select will not work with display: 'none' so instead
                // just banish the input that houses portal URL
                style={{ position: 'absolute', left: '-9999px' }}
              />
              <button
                id="copy"
                className="portal__copy-button"
                value={window.location.href}
                onClick={this.handleCopyURL}
              >
                Copy Portal Link
              </button>
              <button
                className="portal__close-session"
                onClick={this.handleClosePortal}
              >
                Exit Huddle
              </button>
            </div>
          </div>
          {this.props.messages.length > 0 && !this.state.loading && (
            <ul className="portal__message-list" tabIndex="0">
              {messages}
              <li
                style={{ float: 'left', clear: 'both' }}
                ref={el => (this.scrollEl = el)}
              ></li>
            </ul>
          )}
          {this.props.messages.length === 0 && (
            <p className="portal__empty-messages">No messages found.</p>
          )}
          <MessageForm
            handleNewMessage={this.props.handleNewMessage}
            portal_id={this.props.match.params.id}
          />
        </section>
      )
    } else if (
      this.state.gated &&
      !TokenService.hasPortalToken(this.props.match.params.id)
    ) {
      return (
        <form
          className="portal__gate-form"
          onSubmit={this.handlePortalValidation}
        >
          <label htmlFor="validate_password">
            Enter password
            <input
              type="password"
              name="validate_password"
              id="validate_password"
            />
          </label>
        </form>
      )
    } else if (this.state.loading && !this.state.error) {
      return <h2 className="portal__loading-heading">Loading...</h2>
    } else {
      return <h2 className="portal__error">{this.state.error}</h2>
    }
  }
}
