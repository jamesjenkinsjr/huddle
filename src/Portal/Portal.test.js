import React from 'react';
import ReactDOM from 'react-dom';
import Portal from './Portal';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Portal />, div);
  ReactDOM.unmountComponentAtNode(div);
});
