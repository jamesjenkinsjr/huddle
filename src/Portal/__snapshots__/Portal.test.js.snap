// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal component renders Portal by default when not loading 1`] = `
<section
  className="portal__container"
>
  <div
    className="portal__heading-container"
  >
    <div
      className="portal__message-list-heading"
    >
      <h2
        className="portal__title"
      />
      <p
        className="portal__expiry-datetime"
      >
        Expires on
         
        Invalid Date
      </p>
    </div>
    <button
      className="portal__close-session"
      onClick={[Function]}
    >
      Exit Huddle
    </button>
  </div>
  <p
    className="portal__empty-messages"
  >
    No messages found.
  </p>
  <MessageForm
    portal_id=""
  />
</section>
`;

exports[`Portal component renders Portal in loading state by default 1`] = `
<h2
  className="portal__loading-heading"
>
  Loading...
</h2>
`;

exports[`Portal component renders with portal and messages props provided 1`] = `
<section
  className="portal__container"
>
  <div
    className="portal__heading-container"
  >
    <div
      className="portal__message-list-heading"
    >
      <h2
        className="portal__title"
      >
        Test Portal Name
      </h2>
      <p
        className="portal__expiry-datetime"
      >
        Expires on
         
        10/24/2019, 12:09:09 PM
      </p>
    </div>
    <button
      className="portal__close-session"
      onClick={[Function]}
    >
      Exit Huddle
    </button>
  </div>
  <ul
    className="portal__message-list"
    tabIndex="0"
  >
    <Message
      author="Tim"
      content="Hi this is a test"
      create_timestamp="10/23/2019, 12:09:11 PM"
      key="1"
    />
    <Message
      author="Tim"
      content="Hi this is a test"
      create_timestamp="10/23/2019, 12:09:12 PM"
      key="2"
    />
    <Message
      author="Tim"
      content="Hi this is a test"
      create_timestamp="10/23/2019, 12:09:13 PM"
      key="3"
    />
    <li
      style={
        Object {
          "clear": "both",
          "float": "left",
        }
      }
    />
  </ul>
  <MessageForm
    portal_id="a7580f03-d358-4f3f-a5f8-6b69ba02d83a"
  />
</section>
`;
